internal/driverutil/description_test.go

```diff
@@ -0,0 +1,1282 @@
+		}
+
+		if serverDescription.LastUpdateTime != nil {
+			ms := int64(*serverDescription.LastUpdateTime)
+			server.LastUpdateTime = time.Unix(ms/1e3, ms%1e3/1e6)
+		}
+
+		if serverDescription.LastWrite != nil {
+			i, err := strconv.ParseInt(serverDescription.LastWrite.LastWriteDate.Value, 10, 64)
+
+			if err != nil {
+				return err
+			}
+
+			timeWithOffset := baseTime.Add(time.Duration(i) * time.Millisecond)
+			server.LastWriteTime = timeWithOffset
+		}
+
+		if serverDescription.MaxWireVersion != nil {
+			versionRange := NewVersionRange(0, *serverDescription.MaxWireVersion)
+			server.WireVersion = &versionRange
+		}
+
+		if serverDescription.Tags != nil {
+			server.Tags = tag.NewTagSetFromMap(serverDescription.Tags)
+		}
+
+		if test.ReadPreference.MaxStaleness != nil && server.WireVersion == nil {
+			server.WireVersion = &description.VersionRange{Max: 21}
+		}
+
+		servers = append(servers, server)
+	}
+
+	c := description.Topology{
+		Kind:    topologyKindFromString(t, test.TopologyDescription.Type),
+		Servers: servers,
+	}
+
+	if len(test.ReadPreference.Mode) == 0 {
+		test.ReadPreference.Mode = "Primary"
+	}
+
+	readprefMode, err := readpref.ModeFromString(test.ReadPreference.Mode)
+	if err != nil {
+		return err
+	}
+
+	options := make([]readpref.Option, 0, 1)
+
+	tagSets := tag.NewTagSetsFromMaps(test.ReadPreference.TagSets)
+	if anyTagsInSets(tagSets) {
+		options = append(options, readpref.WithTagSets(tagSets...))
+	}
+
+	if test.ReadPreference.MaxStaleness != nil {
+		s := time.Duration(*test.ReadPreference.MaxStaleness) * time.Second
+		options = append(options, readpref.WithMaxStaleness(s))
+	}
+
+	rp, err := readpref.New(readprefMode, options...)
+	if err != nil {
+		return err
+	}
+
+	var selector description.ServerSelector
+
+	selector = &ReadPrefServerSelector{ReadPref: rp}
+	if test.Operation == "write" {
+		selector = &CompositeServerSelector{
+			Selectors: []description.ServerSelector{&WriteServerSelector{}, selector},
+		}
+	}
+
+	result, err := selector.SelectServer(c, c.Servers)
+	if err != nil {
+		return err
+	}
+
+	compareServers(t, test.SuitableServers, result)
+
+	latencySelector := &LatencyServerSelector{Latency: time.Duration(15) * time.Millisecond}
+	selector = &CompositeServerSelector{
+		Selectors: []description.ServerSelector{selector, latencySelector},
+	}
+
+	result, err = selector.SelectServer(c, c.Servers)
+	if err != nil {
+		return err
+	}
+
+	compareServers(t, test.InLatencyWindow, result)
+
+	return nil
+}
+
+func runTest(t *testing.T, testsDir string, directory string, filename string) {
+	filepath := path.Join(testsDir, directory, filename)
+	content, err := ioutil.ReadFile(filepath)
+	require.NoError(t, err)
+
+	// Remove ".json" from filename.
+	filename = filename[:len(filename)-5]
+	testName := directory + "/" + filename + ":"
+
+	t.Run(testName, func(t *testing.T) {
+		var test testCase
+		require.NoError(t, json.Unmarshal(content, &test))
```

Comment:
Optional: Consider using the Extended JSON unmarshaler, since the test format is Extended JSON.

E.g.
```go
require.NoError(t, bson.UnmarshalExtJSON(content, true, &test))
```

Note: We'd probably have to change all of the struct tags to `bson` instead of `json` for that to work properly.